Set up process for sharing code between different components.

ashb opened the pull request:
And move `airflow.utils.timezone` into a shared library as the first example of it working.

In this change we have now setled on an approach using symlinks, but we did
explore other options (see the GH PR for discussion and previous versions,
notably one built upon the `vendoring` tool)

A lot of the reasoning and mode of operation of this is detailed in shared/README.md in this PR, hence why this description is so short.

Currently various places in TaskSDK and Airflow Core both use these utility functions, and while in this specific case they are small enough that they could just be copied and the duplication wouldn't hurt us long term, this changes shows a way in which we can have a single source of truth, but have it included automatically in built dists.

(For posterity, please see https://github.com/apache/airflow/pull/53149/commits/b6ae6a93b82da5338efa5e387a0215fc673c1c28 for the previous vendoring-based approach)

And thanks to Jarek for the help in making the symlink version work, it's much simpler!

kaxil commented on airflow-core/pyproject.toml:
```diff
@@ -203,6 +204,8 @@ YouTube = "https://www.youtube.com/channel/UCSXwxpWZQ7XZ1WL3wqevChA/"
 [tool.hatch.version]
 path = "src/airflow/__init__.py"
 
+[tool.hatch.build.hooks.vendoring]
```
Empty conf?

ashb replied:
Let me double check if it's needed, but I think without it hatch won't use the plugin even if it's installed.

ashb replied:
Yup, needed, I've added a comment here and in task-sdk

kaxil commented on shared/README.md:
```diff
@@ -0,0 +1,54 @@
+
+This folder contains code that is shared across two or more of the Airflow distributions.
+
+## Be Thoughtfull about what you add under here
```
```suggestion
## Be Thoughtful about what you add under here
```

kaxil commented on shared/README.md:
```diff
@@ -0,0 +1,54 @@
+
+### In built distributions
+
+We make use of [hatch-build-time-vendoring] to automatically run [vendoring] for us when the `sdist` is built (and then the wheel is automatically built using that already vendored sdist).
```
I think you meant to add a link here and on vendoring

kaxil replied:
saw it at the botton

potiuk commented on airflow-core/pyproject.toml:
```diff
@@ -26,6 +26,7 @@ requires = [
     "smmap==5.0.2",
     "tomli==2.2.1; python_version < '3.11'",
     "trove-classifiers==2025.5.9.12",
+    "hatch-build-time-vendoring @ https://test-files.pythonhosted.org/packages/54/44/8533528622833807a24261d1fdbc4289a13e124b8ed7461e0d4d963b4ca9/hatch_build_time_vendoring-0.1.dev8-py3-none-any.whl",
```
Should we just include "hatch-build-time-vendoring" in our monorepo instead? That will make any fixes way faster and we will not have to separately release it.

potiuk commented on shared/README.md:
```diff
@@ -0,0 +1,54 @@
+
+One extra thing to note here, is that in order for both IDEs and type checkers to find the "other" copy of the shared library, we commit the typestub files that vendoring generates.
+
+[hatch-build-time-vendoring]: https://pypi.org/project/hatch-build-time-vendoring/
```
I was looking for it and wonder where is the source code for it ? It's not released in PyPI yet - and I wonder if instead relasing it in PyPI we might want to just embed it in the code of Airflow ?

potiuk commented on shared/README.md:
```diff
@@ -0,0 +1,54 @@
+
+`vendoring` is a project from the pip maintainers and is how pip vendors code into it's tree. Now the eagle-eyed amongst you will notice that the project description essentially says "do not use this if you aren't pip", but in this specific case (of vendoring in _our own code_) none of the specific warnings or caveats apply to us.
+
+The main thing the vendoring tool does for us is that it will copy all of the python code out of the source distribution (i.e. `shared/core-and-tasksdk/src/airlfow_shared/*.py`), place it in the target namespace (i.e. `task-sdk/src/airflow/sdk/_vendor/airflow_shared/`) and then most usefully, it will automatically re-write the imports so they use the new names.
```
As discussed in the devlist  discussion - I think it will be better to share specific functionalities under "shared" rather than have one `core-and-task-sdk` - i.e. multiple pyproject.toml files . I am happy to make a follow up PR to do it if we agree that this is a better solution.

ashb replied:
Yes, I hadn't adjusted anything to reflect the latest discussion on the dev list yet, hence why this is still WIP

potiuk commented on shared/core-and-tasksdk/pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+]
+
+
+dependencies = [
```
I think it would be better (following the idea of splitting it "per functionality") - to have separate "sub-project" for each functionality - with it's own dependencies that it needs and then automatically (with pre-commit that we already use for type-stubs) to merge the dependencies into the target projects that are using this functionality.

It could even be done with dependency groups, and include devel dependencies.

Say "apache-airflow-logging" (those are imaginary dependencies):

```pyproject.toml
dependencies = [
   "structlog"
]

[dependency-groups]
dev = [
    "structlog-mock",
    "apache-airflow-devel-common",
]
```

And in the project (say task-sdk) that would use logging that would result in:

```pyproject.toml
dependencies = [
     ....
     # shared logging deps
     "structlog",
     # end of shared logging deps
]

dev = [
     ....
     # shared logging dev deps
     "structlog-mock",
     # end of shared logging dev deps
]
```

This has many advantages:

* maintaining dependencies where they are expected
* uv workspace taking care about conflict resolving and working out if there are no conflicting dependencies
* but most importantly "logging" will become a full locallly developed project on it's own and we could do this:

```bash
cd shared/logging
uv sync
uv run pytest
```

You could **just** open the logging project from Airlow repo and work on it as if it was a completely standalone project - add and run tests etc. and make sure that it has no dependencies to other part of the system (maintaining low cyclomatic complexity).

Then in CI we would just run tests precisely in thiis way as above. The nice thing about `uv sync` in this case is that it will **remove** all other workspace projects and all dependencies that are not used by this "logging" project - and this way we will automatically be sure that we are not using something accidentally from our CI environment and other projects.

ashb replied:
Yes, this is pretty much what I had in mind. I think this layout is independent of how granular or not each of these shared libs actually end up being. But yes, I agree with this.

potiuk commented on task-sdk/src/airflow/sdk/_vendor/.gitignore:
```diff
@@ -0,0 +1 @@
+*/*.py
```
```suggestion
*
```

Should not we ignore **everything** here ? Some of the vendored code might have resources.

potiuk replied:
We should also exclude (!) the `__init__.py`

potiuk replied:
And speaking of "_vendor" name - I think we should not use`_vendor` here as a name - even if we are using `vendoring` tool. Can we configure (mock?) the vendoring tool to use something different (_shared) sounds like a way better name). The `vendor` meaning is very tightly connected with `3rd party` but here we are `vendoring-in` our own code so we are really `1st party` here.

potiuk replied:
It will look better in the imports as well - reflecting it's real use, and if we go for functionality split that will be way more logical: 

```python
from airflow.sdk._shared.logging import logger
```

ashb replied:
Yeah it's already a config setting in pyproject.toml

```toml
[tool.vendoring]
destination = "src/airflow/_vendor/"
requirements = "src/airflow/_vendor/vendor.txt"
namespace = "airflow._vendor"
```

ashb replied:
We do have a quirk we have to avoid here -- if we have `shared/logging/src/logging/__init__.py` this would end up conflicting with the stdlib in dev using this approach. Perhaps another reason to suggest relative imports.

I guess the main issue with relative imports is are we okay with no imports between shared libs? For example, it might be nice to be able to use the timezone/date parsing stuff in the logging one, but I don't think we can do that when symlinking?

potiuk replied:
> I guess the main issue with relative imports is are we okay with no imports between shared libs? For example, it might be nice to be able to use the timezone/date parsing stuff in the logging one, but I don't think we can do that when symlinking?

I think (that was also my cyclomatic complexity arguments) we should avoid it as much as possible - because then it can very easily lead to circular dependencies (config -> logging - > config  that we currently have is a classic example of it). Most of this can be solved by injecting stuff at the higher level - for example during explicit initialization. It makes the "acyclic tree" hierarchy of dependencies between such shared code, which is precisely what low cyclomatic complexity is about. 

So in a way - adding "friction" for the shared code to use another shared code is likely a good thing. You should **not** be able to use the shared code from another module to easily. But also - it's not impossible in case of symlinks - you can have symlinks from one shared library to the other. That will work exactly the same way as symlinking shared library in the main code. And I think it's the right "amount of friction" - because you not only have to import stuff from the other library, but have to make symlink and it will - much more obviously - raise eyebrows during review - and cases like "logging uses config and config uses logging" is somethign that will be definitely "no go" as it will "visibly" create circular dependencies.

ashb committed:
Set up process for sharing code between different components

And move `airflow.utils.timezone` into a shared library as the first example
of it working.

In this change we have now setled on an approach using symlinks, but we did
explore other options (see the GH PR for discussion and previous versions,
notably one built upon the `vendoring` tool)

A lot of the reasoning and mode of operation of this is detailed in
shared/README.md in this PR, hence why this description is so short.

Currently various places in TaskSDK and Airflow Core both use these utility
functions, and while in this specific case they are small enough that they
could just be copied and the duplication wouldn't hurt us long term, this
changes shows a way in which we can have a single source of truth, but have it
included automatically in built dists.

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>
Co-authored-by: Amogh Desai <amoghrajesh1999@gmail.com>

amoghrajesh commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+
+## Mechanics of sharing
+
+The primary mechanism by which we share code is to use in-repo symlinks. This means that there is a single copy of the code exists in the repo (no need for a pre-commit to update other copies, resulting in larger diffs).
```
```suggestion
The primary mechanism by which we share code is to use in-repo symlinks. This means that there is a single copy of the code existing in the repo (no need for a pre-commit to update other copies, resulting in larger diffs).
```

amoghrajesh commented on shared/README.md:
````diff
@@ -0,0 +1,91 @@
+from ..timezones.timezone import is_naive  # noqa: TID252
+```
+
+This is different to the rest of the airflow codebase where relative imports are not allowed.
````
This part could have a precommit.

amoghrajesh replied:
Maybe in a follow up.

potiuk replied:
Yes. I am planning to add a pre-commit that will look at what distribution uses what shared code (and what shared code uses other shared code) and synchronize the pyproject.toml definitions and imports (very similar pre-commit to one we have in providers - where we are doing similar walk through pyproject.toml, provider.yamls __init__.py and verify that all of them are  in sync and dependencies are "correct" - this one will be part of the check.

amoghrajesh commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+
+### Shared libraries should be grouped one level beneath `airflow_shared/`
+
+In order to make both the relative imports above function and to provider simple structure, we want to arrange things in single named chunks, where the folder name is also the "shared library name". For example
```
```suggestion
In order to make both the relative imports above function and to provide simple structure, we want to arrange things in single named chunks, where the folder name is also the "shared library name". For example
```

amoghrajesh commented on task-sdk/src/airflow/sdk/timezone.py:
```diff
@@ -0,0 +1,49 @@
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
```
The public api facade looks nice too.

potiuk replied:
I like it too - especially if paired with "no import of those from _shared" ruff rule. that will make our code way, way cleaner. Good idea @ashb

potiuk replied:
Question is -> do we want to  (eventually) automate it as well eventually in all distributions including the shared libraries, and how. It would be great for example that when shared library is included in a few distributions, we do not have to remember about exposing newly added public methods everywhere where the library is included.

I see  several options:

1) document the pattern in the "shared" approach documentation and let people do it manually (prone to having different set of public stuff in different distributions)

2) autamatically expose everything (that's a bit against the comment below - probably we do not want to expose everything

3) define some conventions in shared libraries to say "this is public, this is not" and generate those facades automatically based on that in pre-commit.

4) add this code to `__init__.py` of the shared library and have pre-commit that copies it to all libraries where it is used and replaces the import.

WDYT?

potiuk replied:
(all that can be done in the follow-up).

kaxil replied:
https://airflow.apache.org/docs/task-sdk/stable/api.html -- Anything in here becomes public. So we can add things over there, systematically.

potiuk replied:
My thinking was that the facade is not only for task-sdk, but we could also use it in "airflow-core" - so that we do no have to use "from airflow.sdk impor timezone" and "from airflow import timezone" in core - but yeah now when I think of that it should only be in "task-sdk" and indeed serve as e public api, so maintaining it manually is actually best approach.

uranusjr reviewed (approved):
Have we discussed backward compatibility (in general)? Like should `airflow.utils.timezone` should still be accessible by users.

uranusjr commented on task-sdk/tests/task_sdk/execution_time/test_task_runner.py:
```diff
@@ -1130,7 +1130,7 @@ def test_get_context_without_ti_context_from_server(self, mocked_parse, make_ti_
     def test_get_context_with_ti_context_from_server(self, create_runtime_ti, mock_supervisor_comms):
         """Test the context keys are added when sent from API server (mocked)"""
-        from airflow.utils import timezone
+        from airflow.sdk import timezone
```
Why do we need to re-import in this test? `timezone` is already imported at the top level. Is this testing something specific?

potiuk replied:
Looks like a left-over :)

ashb replied:
Yeah, I just did this with a find-and-replace, so it was already double imported. I'll fix it though.

amoghrajesh commented on shared/timezones/src/airflow_shared/timezones/timezone.py:
```diff
@@ -0,0 +1,318 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
```
Should we be adding an entry to deprecate the `airflow.utils.timezone` in `airflow-core/src/airflow/utils/__init__.py`?

potiuk replied:
That's a good point - the question is "where" it should live and which versions, because (eventually) different components migh have diffferent distributions installed.

1) On worker we are going to have only "apache-airlflow-task-sdk" installed (no "apache-airflow", no "apache-airflow-core") on day one of Airflow 3.1 (hopefully)

2) In other components (dag file processor, triggerer, and api-server if we want to continue installing providers there) - we will have both:

* "apache-airflow-task-sdk"
* "apache-airflow-core"

3) On scheduler on the other hand only "apache-airflow-core" - no "apache-airflow-task-sdk" (not for now - this will only be possible if we separate executors from "regular" providers which might be next frontier of the separation) 

Theorethically of course - many people will just have "apache-airflow" installed and they will have "everything".

So ... the queston is where the "utils" deprecation code should live.

I think we do not need it in in "core" because we have full control over it and no legacy "user" code is allowed , so likely it could live in "task-sdk" - and "apache-airflow-task-sdk" should also provide "airflow.utils" package with all deprecations (redirectiong to _shared libraries of task-sdk"

ashb replied:
For now/for 3.1 the deprecation should go in airflow-core, as that's where the `airflow.utils` module lives, but once (if?) we re-organize airflow-core then it then compat layers should move into task-sdk, I agree

amoghrajesh reviewed (approved):
Few nits and a comment about backward compatibility.

Rest looks nice!

potiuk commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+
+The primary reason we are using this sharing approach, rather than the perhaps more traditional approach of depending on a shared distribution from PyPI is around compatibility. Lets say that we have a library to parse AirflowConfig (including all the sources, env, `_cmd` handling, etc etc.). If that code is used by two distributions that are installed in the same Python environment, then either  ...
+
+Instead by using an approach similar to vendoring, where each dist ships with a copy of the code of the version it know it works with it allows them to co-exist in the same env without problems, and also reduces the cognitive load on developers when making changes to the shared libs -- as long as it passes the CI it should be good (and we don't have to test an ever increasing matrix of versions t ...
```
```suggestion
Instead by using an approach similar to vendoring (or - if you compare it to C libraries - static linking), where each dist ships with a copy of the code of the version it know it works with it allows them to co-exist in the same env without problems, and also reduces the cognitive load on developers when making changes to the shared libs -- as long as it passes the CI it should be good (and we don't have to test an ever increasing matrix of versions to ensure compatibility.)
```

kaxil replied:
```suggestion
Instead by using an approach similar to vendoring, where each dist ships with a copy of the code of the version it knows it works with, it allows them to co-exist in the same env without problems, and also reduces the cognitive load on developers when making changes to the shared libs -- as long as it passes the CI it should be good (and we don't have to test an ever increasing matrix of versions to ensure compatibility.)
```

potiuk commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+ under the License.
+ -->
+
+# Shared Python Code for Airflow Components
```
I ❤️ the documentation!

amoghrajesh replied:
+1

potiuk commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+## Use In Editable installs/Git checkouts
+
+In the editable checkout (which is what you get by default when using `uv sync` etc), the shared libraries are symlinked directly in place, for example `task-sdk/src/airflow/sdk/_shared/timezones` is symlinked to `shared/timezones/src/airflow_shared/timezones`. This means that python automatically finds the files underneath the module as if they lived directly under `airflow.sdk._shared.timezones ...
+
```
```suggestion
Also, the distribution that have library symlinked, should have dependencies added that the shared library needs. TODO(potiuk): will add pre-commit that will synchronize and check if the dependencies are properly set, so you do not have to remember about it.
```

potiuk commented on shared/README.md:
````diff
@@ -0,0 +1,91 @@
+"../shared/timezones/src/airflow_shared/timezones" = "src/airflow/sdk/_shared/timezones"
+```
+
+We will need one entry in that table for each module we include directly below `src/airflow/sdk/_shared`.
````
```suggestion
We will need one entry in that table for each module we include directly below `src/airflow/sdk/_shared`.
TODO(potiuk): will add pre-commit very soon to synchronize and check it automatically.
```

potiuk commented on shared/timezones/pyproject.toml:
```diff
@@ -0,0 +1,47 @@
+[project]
+name = "apache-airflow-shared-timezones"
+description = "Shared timezone code for Airflow distributions"
+version = "0.0"
```
I guess it would be great to figure out automated versioning scheme and automatically bump the version here. It would really be good to ba able to know which `version` has been used in which _shared "includie". I think we can use the built-in dynamic versioning feature of build tooling for that, or maybe better - every time when code of the shared library change, this could be bumped automatically - I will think about it and add it in the pre-commit follow-up.

potiuk replied:
Of course the **right** version wll be checked out when you check-out the tag for distribution using it, so this is not **as** important, but if we have a consistent. incremental scheme here, that might be good to see "ah ... the thing was added in this and that version". We might even get automatically generated mini-changelog showing which version added which features. Might be really important to reason about some problems.

ashb replied:
Versioning is irrelevant until we choose to release it. We can tackle that when if/we decide to release these on pypi.

The entire point of symlinking these in as there is no versioning to worry about.

potiuk replied:
I think it's more of "tracking" that versioning. But yes we do not strictly need to "solve it" now.

potiuk commented on shared/timezones/pyproject.toml:
```diff
@@ -0,0 +1,47 @@
+build-backend = "hatchling.build"
+
+[tool.hatch.build.targets.wheel]
+packages = ["src/airflow_shared"]
```
```suggestion
packages = ["src/airflow_shared/timezones"]
```

or 

```suggestion
packages = ["src/airflow_shared", "src/airflow_shared/timzeones"]
```
I am not entirely sure what are the consequences of using the "packages" entry here in wheel ?

ashb replied:
It didn't seem to affect anything and worked as it was? (I tested the `uv build shared/timezones` and then `unzip -l` to see what it put in it)

potiuk replied:
Ah indeed -> https://hatch.pypa.io/latest/config/build/#explicit-selection . This is only to aid explicit selection of what should be in the package, overriding default auto-detection. So "airflow_shared" in this case is enough (because there is nothing else than the right sub-package there):

```
Generic[¶](https://hatch.pypa.io/latest/config/build/#generic)
You can use the only-include option to prevent directory traversal starting at the project root and only select specific relative paths to directories or files. Using this option ignores any defined [include patterns](https://hatch.pypa.io/latest/config/build/#patterns).


[ pyproject.toml](https://hatch.pypa.io/latest/config/build/#pyprojecttoml_5)
[ hatch.toml](https://hatch.pypa.io/latest/config/build/#hatchtoml_4)

[tool.hatch.build.targets.sdist]
only-include = ["pkg", "tests/unit"]

Packages[¶](https://hatch.pypa.io/latest/config/build/#packages)
The packages option is semantically equivalent to only-include (which takes precedence) except that the shipped path will be collapsed to only include the final component.

So for example, if you want to ship a package foo that is stored in a directory src you would do:


[ pyproject.toml](https://hatch.pypa.io/latest/config/build/#pyprojecttoml_6)
[ hatch.toml](https://hatch.pypa.io/latest/config/build/#hatchtoml_5)

[tool.hatch.build.targets.wheel]
packages = ["src/foo"]
```

potiuk commented on shared/README.md:
````diff
@@ -0,0 +1,91 @@
+    │           ├── __init__.py
+    └── tests/
+```
+
````
Maybe we should add 

> It is important that the shared library `pyproject.toml` has the following entries:

```pyproject.toml
classifiers = [
    "Private :: Do Not Upload",
]
[tool.ruff]
extend = "../../pyproject.toml"
src = ["src"]

[tool.ruff.lint.per-file-ignores]
# Ignore Doc rules et al for anything outside of tests
"!src/*" = ["D", "S101", "TRY002"]

[tool.ruff.lint.flake8-tidy-imports]
# Override the workspace level default
ban-relative-imports = "parents"
```

=
or maybe better tell "Copy the base pyproject.toml from existing shared library and modify distribution name, package, description (might be more future proof).

kaxil commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+
+### Shared libraries _must_ use relative imports for other shared libraries**
+
+The one caveat to this is is due to the side-effect that these shared modules are going to be imported from different locations (for example `airflow._shared.timezones.timezone` and `airflow.sdk._shared.timezones.timezone`) then any imports inside the shared code to other parts of shared libraries _must_ make use of relative imports, and possibly needs `# noqa: TID252`
```
```suggestion
The one caveat to this is due to the side-effect that these shared modules are going to be imported from different locations (for example `airflow._shared.timezones.timezone` and `airflow.sdk._shared.timezones.timezone`) then any imports inside the shared code to other parts of shared libraries _must_ make use of relative imports, and possibly needs `# noqa: TID252`
```

potiuk reviewed (approved):
Looks fantastic! It's great we could combine all the ideas and get something really nice and developer-friendly :) . There are few comments - most of them require more a of discussion, agreement and follow-up than changing the PR.

kaxil commented on shared/README.md:
```diff
@@ -0,0 +1,91 @@
+
+### Including In built distributions
+
+Symlinks will not work when we build the main distributions for PyPI and other channels (since the symlinks would be to files outside of the distribution otherwise).
```
```suggestion
Symlinks will not work when we build the main distributions for PyPI and other channels (since the symlinks would be to files outside of the distribution otherwise). This is specific to sdist; building wheel resolves the symlinks.
```

kaxil commented on airflow-core/src/airflow/utils/db_cleanup.py:
```diff
@@ -36,10 +36,10 @@
 from airflow.cli.simple_table import AirflowConsole
 from airflow.configuration import conf
 from airflow.exceptions import AirflowException
-from airflow.utils import timezone
```
We should nuke `airflow-core/src/airflow/utils/timezone.py` too since it is no longer used I assume, and add a back-compat entry in `airflow-core/src/airflow/utils/__init__.py`

ashb replied:
Oh yes, I forgot to do that in this reworked PR. On it.

ashb commented on task-sdk/src/airflow/sdk/timezone.py:
```diff
@@ -0,0 +1,49 @@
+    "parse",
+    "utc_epoch",
+    "utcnow",
+]
```
Now I think about it, I wonder if I could remove some of these from the public API -- or do we want to all of these?

I'm not sure `make_naive`,`make_aware` are needed.

```suggestion
__all__ = [
    "coerce_datetime",
    "convert_to_utc",
    "datetime",
    "from_timestamp",
    "local_timezone",
    "parse",
    "utc_epoch",
    "utcnow",
]
```

WYDT @kaxil @amoghrajesh @potiuk ?

kaxil replied:
Yeah, that's sensible. `parse` and `datetime` are the ones that are most used in DAGs from what I have seen.

ashb replied:
Yeah, some of these were already marked with `:meta private:` so I've removed those, and `utc_eopch` was used in exactly one place (dag processor) so I moved it to there.
